### PR TITLE
Reconcile dataplane

### PR DESCRIPTION
**Pull request by fao89**
copied from: https://github.com/openstack-k8s-operators/openstack-operator/pull/230

### CODE REVIEW

File: controllers/openstackdataplanenode_controller.go [hunk 1]
```diff
@@ -138,16 +138,15 @@ func (r *OpenStackDataPlaneNodeReconciler) Reconcile(ctx context.Context, req ct
...
-			instance.Status.Conditions.MarkTrue(condition.ReadyCondition, dataplanev1beta1.DataPlaneNodeReadyMessage)
-		}
-		c := instance.Status.Conditions.Mirror(dataplanev1beta1.DataPlaneNodeReadyCondition)
-		if c.Reason == condition.ErrorReason {
-			instance.Status.Conditions.MarkFalse(
-				condition.ReadyCondition,
-				condition.ErrorReason,
-				condition.SeverityError,
-				c.Message)
+			instance.Status.Conditions.MarkTrue(condition.ReadyCondition, condition.ReadyMessage)
```
Comment (slagle) [inline controllers/openstackdataplanenode_controller.go]: the message here should be condition.DataPlaneNodeReadyMessage

File: controllers/openstackdataplane_controller.go [hunk 2]
```diff
@@ -62,5 +218,97 @@ func (r *OpenStackDataPlaneReconciler) Reconcile(ctx context.Context, req ctrl.R
...
+		err := client.Get(ctx, types.NamespacedName{Name: node.Name, Namespace: instance.Namespace}, node)
+		if err != nil && !k8s_errors.IsNotFound(err) {
+			logger.Info("Failed to get Node", "Node.Namespace", instance.Namespace, "Node.Name", node.Name)
+			return err
+		}
+		if k8s_errors.IsNotFound(err) {
+			err := client.Create(ctx, node)
+			if err != nil {
+				logger.Info("Failed to create Node", "Node.Namespace", instance.Namespace, "Node.Name", node.Name)
+				return err
```
Comment (slagle) [inline controllers/openstackdataplane_controller.go:276]: I am getting an error here:

'''
2023-03-17T13:02:34-04:00       INFO    controllers.OpenStackDataPlane  Failed to create Node   {"Node.Namespace": "openstack", "Node.Name": "edpm-compute-1"}               2023-03-17T13:02:38-04:00       ERROR   Reconciler error        {"controller": "openstackdataplane", "controllerGroup": "dataplane.openstack.org", "controllerKind": "OpenStackDataPlane", "OpenStackDataPlane": {"name":"openstack-edpm","namespace":"openstack"}, "namespace": "openstack", "name": "openstack-edpm", "reconcileID": "dbe7510e-f1fb-4b5b
-ae71-f8c97aa2b3d8", "error": "resourceVersion should not be set on objects to be created"}                                                                                  sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /home/jslagle/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:329                                                            
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem                                                                                             /home/jslagle/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:274
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2                                                                                           
        /home/jslagle/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:235                                                            2023-03-17T13:02:38-04:00       INFO    controllers.OpenStackDataPlane  CreateDataPlaneResources
'''

File: config/samples/dataplane_v1beta1_openstackdataplane.yaml [hunk 3]
```diff
@@ -1,7 +1,50 @@
 apiVersion: dataplane.openstack.org/v1beta1
 kind: OpenStackDataPlane
 metadata:
-  name: openstackdataplane-sample
+  name: openstack-edpm
 spec:
-  dataPlaneRoles:
-    - openstackdataplanerole-sample
+  deployStrategy:
+      deploy: true
```
Comment (slagle) [inline config/samples/dataplane_v1beta1_openstackdataplane.yaml:7]: can we set this to false here to match the other samples?

File: controllers/openstackdataplane_controller.go [hunk 4]
```diff
@@ -50,10 +60,156 @@ type OpenStackDataPlaneReconciler struct {
...
+		instance.InitConditions()
+		// Register overall status immediately to have an early feedback e.g. in the cli
+		return ctrl.Result{}, nil
+	}
+
+	// Reset all ReadyConditons to 'Unknown'
+	instance.InitConditions()
+
+	// If we're not deleting this and the service object doesn't have our finalizer, add it.
+	if instance.DeletionTimestamp.IsZero() && controllerutil.AddFinalizer(instance, helper.GetFinalizer()) {
```
Comment (slagle) [inline controllers/openstackdataplane_controller.go:124]: It looks like the finalizer causes the delete to just hang indefinitely. Don't we need to implement the finalizer if we add it? Do we need this line at all?

File: config/samples/dataplane_v1beta1_openstackdataplane.yaml (hunk 3 above)
Comment (fao89) [reply config/samples/dataplane_v1beta1_openstackdataplane.yaml:7]: yes

File: controllers/openstackdataplane_controller.go (hunk 4 above)
Comment (fao89) [reply controllers/openstackdataplane_controller.go:124]: I'll remove it

File: controllers/openstackdataplane_controller.go [hunk 5]
```diff
@@ -62,5 +218,97 @@ func (r *OpenStackDataPlaneReconciler) Reconcile(ctx context.Context, req ctrl.R
...
+
+// CreateDataPlaneNode -
+func CreateDataPlaneNode(ctx context.Context, instance *dataplanev1beta1.OpenStackDataPlane, helper *helper.Helper) error {
+	client := helper.GetClient()
+	logger := helper.GetLogger()
+	node := &dataplanev1beta1.OpenStackDataPlaneNode{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: instance.Namespace,
+		},
+	}
```
Comment (slagle) [inline controllers/openstackdataplane_controller.go:262]: You need to move this node assignment into the loop below. What is causing the traceback is that it creates the first node, which sets a resourceVersion, then when it goes to create the 2nd node, it reuses this same var where resourceVersion is set.

File: controllers/openstackdataplane_controller.go [hunk 6]
```diff
@@ -50,10 +60,146 @@ type OpenStackDataPlaneReconciler struct {
...
+			return ctrl.Result{}, err
+		}
+
+		instance.Status.Conditions.Set(condition.FalseCondition(condition.ReadyCondition, condition.InitReason, condition.SeverityInfo, dataplanev1beta1.DataPlaneRoleReadyWaitingMessage))
+		for _, role := range roles.Items {
+			if role.Spec.DataPlane != instance.Name {
+				err = fmt.Errorf("role %s: role.DataPlane does not match with role.Label", role.Name)
+				deployErrors = append(deployErrors, "role.Name: "+role.Name+" error: "+err.Error())
+			}
+			role.Spec.DeployStrategy.Deploy = instance.Spec.DeployStrategy.Deploy
```
Comment (slagle) [inline controllers/openstackdataplane_controller.go:158]: You need to move this line inside the CreateOrPatch func, otherwise it doesn't take effect.

**Comment by slagle** [comment]
+2

**Pull request merged**